### PR TITLE
RDM-11411: Code changes required for the Create functionality

### PR DESCRIPTION
**Pull request by kiran-yenigala-hmcts**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-11411

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ Y] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/service/casedataaccesscontrol/RoleAssignmentsFilteringServiceImpl.java [hunk 1]
```diff
@@ -28,62 +24,33 @@ public RoleAssignmentsFilteringServiceImpl(List<RoleAttributeMatcher> roleAttrib
...
-                Pair<RoleAssignment, RoleMatchingResult> resultPair = Pair.of(roleAssignment, new RoleMatchingResult());
-
-                roleAttributeMatchers.forEach(matcher -> matcher.matchAttribute(resultPair, (CaseTypeDefinition) null));
-                return resultPair;
-            })
-            .filter(resultPair -> resultPair.getRight().matchedAllValues())
+            .filter(roleAssignment -> roleAttributeMatchers
+                .stream()
+                .map(matcher -> matcher.matchAttribute(roleAssignment, caseDetails))
+                .allMatch(matched -> matched == true))
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/casedataaccesscontrol/RoleAssignmentsFilteringServiceImpl.java:37]:
```suggestion
                .allMatch(matcher -> matcher.matchAttribute(roleAssignment, caseDetails)))
```
And same simplification for the other methods

Comment (danlysiak) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/casedataaccesscontrol/RoleAssignmentsFilteringServiceImpl.java:37]: Could also consider a common method for some of the repeated code.. e.g. introduce a new method
```
    private List<RoleAssignment> filterMatchingRoleAssignments(RoleAssignments roleAssignments, BiPredicate<RoleAttributeMatcher, RoleAssignment> hasMatch) {
        return roleAssignments
            .getRoleAssignments()
            .stream()
            .filter(roleAssignment -> roleAttributeMatchers
                .stream()
                .allMatch(matcher -> hasMatch.test(matcher, roleAssignment)))
            .collect(Collectors.toList());
    }
``` 
which would turn this into:
```
return filterMatchingRoleAssignments(roleAssignments,
            (matcher, roleAssignment) -> matcher.matchAttribute(roleAssignment, caseDetails));
```

File: src/main/java/uk/gov/hmcts/ccd/domain/service/AuthorisationMapper.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            && !authorisations.isEmpty()) {
+            Collection<String> filterAuthorisations = CollectionUtils
+                .intersection(roleAssignmentAuthorisations, authorisations);
+
+            return !filterAuthorisations.isEmpty();
+        }
+        return authorisations.isEmpty();
+    }
+
+    default List<AccessProfile> createAccessProfiles(RoleAssignment roleAssignment,
```
Comment (rafalkalita) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/AuthorisationMapper.java:37]: Not used I think.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/AuthorisationMapper.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang.BooleanUtils;
+import uk.gov.hmcts.ccd.domain.model.casedataaccesscontrol.AccessProfile;
+import uk.gov.hmcts.ccd.domain.model.casedataaccesscontrol.RoleAssignment;
+import uk.gov.hmcts.ccd.domain.model.definition.RoleToAccessProfileDefinition;
+
+public interface AuthorisationMapper {
```
Comment (rafalkalita) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/AuthorisationMapper.java:14]: Why an interface with the default methods and not a regular bean?

**Review by rafalkalita — COMMENTED** [review]
2 comments

**Review by danlysiak — COMMENTED** [review]
The extra refactoring looks a lot cleaner 👌 A few comments..

File: src/main/java/uk/gov/hmcts/ccd/domain/service/AuthorisationMapper.java (hunk 2 above)
Comment (rafalkalita) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/AuthorisationMapper.java:37]: True. Haven't seen it, because the code was folded :)

**Review by rafalkalita — APPROVED** [review]

**Pull request merged**